core/primitives/src/time.rs

```diff
@@ -82,74 +74,107 @@ impl Drop for MockClockGuard {
                         panic!("Mock clock run out of samples");
                     }
                 }
-            } else {
-                chrono::Utc::now()
             }
+            None => chrono::Utc::now(),
         })
     }
+}
 
-    pub fn instant() -> Instant {
-        MockClockPerThread::with(|clock| {
-            if clock.is_mock {
-                let x = clock.pop_instant();
-                match x {
-                    Some(t) => t,
-                    None => {
-                        panic!("Mock clock run out of samples");
-                    }
-                }
-            } else {
-                Instant::now()
-            }
-        })
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    #[should_panic]
+    fn test_clock_panic_utc() {
+        let _mock_clock_guard = MockClockGuard::default();
+        Clock::utc();
     }
 
-    pub fn instant_call_count() -> u64 {
-        MockClockPerThread::with(|clock| clock.instant_call_count)
+    #[test]
+    #[should_panic]
+    fn test_clock_panic_instant() {
+        let _mock_clock_guard = MockClockGuard::default();
+        Clock::instant();
     }
 
-    pub fn utc_call_count() -> u64 {
-        MockClockPerThread::with(|clock| clock.utc_call_count)
+    #[test]
+    fn test_clock() {
+        {
+            let mock_clock_guard = MockClockGuard::default();
+
+            let utc_now = Utc::now();
+            mock_clock_guard.add_utc(
+                utc_now
+                    .checked_add_signed(chrono::Duration::from_std(Duration::from_secs(1)).unwrap())
+                    .unwrap(),
+            );
+            mock_clock_guard.add_utc(
+                utc_now
+                    .checked_add_signed(chrono::Duration::from_std(Duration::from_secs(2)).unwrap())
+                    .unwrap(),
+            );
+            mock_clock_guard.add_utc(
+                utc_now
+                    .checked_add_signed(chrono::Duration::from_std(Duration::from_secs(3)).unwrap())
+                    .unwrap(),
+            );
+            assert_eq!(
+                Clock::utc(),
+                utc_now
+                    .checked_add_signed(chrono::Duration::from_std(Duration::from_secs(1)).unwrap())
+                    .unwrap(),
+            );
+            assert_eq!(
+                Clock::utc(),
+                utc_now
+                    .checked_add_signed(chrono::Duration::from_std(Duration::from_secs(2)).unwrap())
+                    .unwrap(),
+            );
+            assert_eq!(
+                Clock::utc(),
+                utc_now
+                    .checked_add_signed(chrono::Duration::from_std(Duration::from_secs(3)).unwrap())
+                    .unwrap(),
+            );
+
+            assert_eq!(mock_clock_guard.utc_call_count(), 3);
+            drop(mock_clock_guard);
+        }
+        let mock_clock_guard = MockClockGuard::default();
+        assert_eq!(mock_clock_guard.utc_call_count(), 0);
     }
```

Comment:
can you also add a simple test that checks threading behavior?

Reply:
I added a test.